### PR TITLE
:books: [#42] add example use to docs

### LINKED ISSUE #42: As a developer I want examples how to use the lib
### Thema / Theme

Other

### Description

* [x] Please update "Usage" section of the README file with examples how to use the library for import and export
* [x] Please add more information in the docstrings of `get_sharing_configs_import_data` and `get_sharing_configs_export_data` 

### Example

As example we're going to use "themes", that changes the appearance of the Django admin.
For simplicity, a color theme has just a few (css) properties and a name. This "theme" can be exported and imported via this library.

* [x] Add a "Theme" model, with a fields: `name, primary, accent, secondary, primary-fg`. You can use [Django Colorfield](https://pypi.org/project/django-colorfield/)
* [x] Add a "Configuration" model (can be [django-solo](https://github.com/lazybird/django-solo) model but doesn't have to) with 1 dropdown field: `theme`
* [x] The selected theme in the configuration, should be loaded in the Django admin `base_site.html` template. This can be done with a Django context processor to load in the theme (for simplicity) and with some inline `<style>` to set the theme (see screenshots: `:root`)

We can accomplish this with minimal effort:

**Theme settings:**
image

**Themes:**
image
image

### Added value

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by frekkenSnork**
Fixes #42
### **update from the last commit 10:30 June 15:**
After rebase  and solving conflicts got failed Codecov test. Couldn't see all details (not enough authorized). 
Could you provide me with more info why this test failed. Thanks.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -109,4 +109,7 @@ venv.bak/
 # My own files/folders
 
 mock_error.py
-mock_data
\ No newline at end of file
+mock_data
+manage.py
+testapp/sharing_configs.db
```
Comment (joeribekker) [inline .gitignore:114]: `testapp/sharing_configs.db` This can be gone. The line below will cover it.

File: README.rst [hunk 2]
```diff
@@ -87,4 +87,113 @@ the exporting file and override ``get_sharing_configs_export_data`` method
     :target: https://pypi.org/project/sharing_configs/
 
 .. _Maykin Media B.V.: https://www.maykinmedia.nl
-.. _Sharing Configs API: https://github.com/maykinmedia/sharing-configs-api.git
\ No newline at end of file
+.. _Sharing Configs API: https://github.com/maykinmedia/sharing-configs-api.git
+
+Example
+=======
+Simple example for changing color theme in the django admin:
```
Comment (joeribekker) [inline README.rst:94]: The example should focus on the sharing aspect. Now, the example focuses on the color switching. So, let's rewrite it a little bit:

```suggestion

We can use the Sharing Configs library to exchange color-themes for the Django admin with other users. For this, we need a model that stores the color-theme, and use the Sharing Configs mixins to import and export the color-theme.
```

File: README.rst [hunk 3]
```diff
@@ -87,4 +87,113 @@ the exporting file and override ``get_sharing_configs_export_data`` method
...
+.. code-block:: python
+
+    from django.db import models
+    from django.utils.translation import gettext_lazy as _
+    from solo.models import SingletonModel
+
+
+    class Configuration(SingletonModel):
+        """
+        configuration for export/import of color theme in the Django admin
```
Comment (joeribekker) [inline README.rst:107]:
```suggestion
        Configuration that holds the current theme.
```

File: README.rst [hunk 4]
```diff
@@ -87,4 +87,113 @@ the exporting file and override ``get_sharing_configs_export_data`` method
...
+            convert byte string into a dictionary and create a color-theme object;           
+            example api response: b'{"accent": "#f8f8f8", "name": "some-name", "primary": "#ab8585", "primary_fg": "#1a2b3c", "secondary": "#315980"}'
+            """              
+            decoded_content = content.decode("utf-8")
+            theme_dict = json.loads(decoded_content)        
+            return ColorTheme.objects.create(**theme_dict)       
+
+
+Create a ``context_processors.py`` file.  
+Add ``yourapp.context_processors.set_color`` to the list of TEMPLATES context_processors in ``settings.py``
```
Comment (joeribekker) [inline README.rst:169]:
```suggestion
That takes care of the import and export functionality for exchaning color-themes. To make it actually working, we complete this example with some additional code.

Create a ``context_processors.py`` file, to pass the currently configured theme to the template context:
```

File: README.rst [hunk 5]
```diff
@@ -87,4 +87,113 @@ the exporting file and override ``get_sharing_configs_export_data`` method
...
+        secondary_color = color_theme_obj.secondary
+        accent = color_theme_obj.accent
+        primary_fg = color_theme_obj.primary_fg    
+        return {
+            "primary": primary_color,
+            "secondary": secondary_color,            
+            "primary_fg": primary_fg,
+            "accent": accent,
+            
+        }
```
Comment (joeribekker) [inline README.rst:189]: Let's make this a little bit more compact. Also the FK-relation is weird to read like this:

```suggestion
        theme = conf.theme
        if theme is None:
            return {}
       
        return {
            "primary": theme.primary,
            "secondary": theme.secondary,            
            "primary_fg": theme.primary_fg,
            "accent": theme.accent,
        }
```

File: sharing_configs/utils.py [hunk 6]
```diff
@@ -19,7 +19,7 @@ def get_folders_from_api(permission: Optional[str]) -> dict:
 
         return data
     except (requests.exceptions.HTTPError, requests.ConnectionError) as exc:
-        raise ApiException({"error": "No folders"})
+        raise ApiException("No folders")
```
Comment (joeribekker) [inline sharing_configs/utils.py:22]:
```suggestion
        raise ApiException("Could not retrieve any folders due to a connection error.")
```

File: sharing_configs/client_util.py [hunk 7]
```diff
@@ -39,43 +40,49 @@ def export(self, folder, data) -> dict:
...
+                    params=permission,
+                )
+            else:
+                resp = requests.get(
+                    url=self.get_list_folders_url(), headers=self.headers
+                )
             resp.raise_for_status()
         except (requests.exceptions.HTTPError, requests.ConnectionError) as exc:
-            raise ApiException({"error": "No folders"})
+            raise ApiException("No folders available")
```
Comment (joeribekker) [inline sharing_configs/client_util.py:85]:
```suggestion
            raise ApiException("Could not retrieve any folders due to a connection error.")
```

File: sharing_configs/client_util.py [hunk 8]
```diff
@@ -39,43 +40,49 @@ def export(self, folder, data) -> dict:
...
         try:
+
+            resp = requests.get(
+                url=self.get_import_url(folder, filename), headers=self.headers
+            )
+
             resp.raise_for_status()
         except (requests.exceptions.HTTPError, requests.ConnectionError) as e:
+
             raise ApiException("Error during import of object")
```
Comment (joeribekker) [inline sharing_configs/client_util.py:65]:
```suggestion
            raise ApiException("Could not import the item due to a connection error.")
```

File: sharing_configs/client_util.py [hunk 9]
```diff
@@ -39,43 +40,49 @@ def export(self, folder, data) -> dict:
...
-            url=self.get_export_url(folder), headers=self.headers, json=data
-        )
+
         try:
+            resp = requests.post(
+                url=self.get_export_url(folder), headers=self.headers, json=data
+            )
             resp.raise_for_status()
         except (requests.exceptions.HTTPError, requests.ConnectionError) as e:
             raise ApiException("Error during export of object")
```
Comment (joeribekker) [inline sharing_configs/client_util.py:50]:
```suggestion
            raise ApiException("Could not export the item due to a connection error.")
```

File: sharing_configs/admin.py [hunk 10]
```diff
@@ -187,25 +186,29 @@ def import_from_view(self, request, extra_context=None):
...
                 filename = form.cleaned_data.get("file_name")
-                obj = SharingConfigsClient()
+                obj_client = SharingConfigsClient()
                 try:
-                    content = obj.import_data(folder, filename)
+                    content = obj_client.import_data(folder, filename)
                     obj = self.get_sharing_configs_import_data(content)
                     msg = format_html(
-                        _(f"The (file) object {obj} has been imported successfully!"),
+                        _("The (file) object {object} has been imported successfully!"),
```
Comment (joeribekker) [inline sharing_configs/admin.py:197]:
```suggestion
                        _("The item {object} has been imported successfully!"),
```

File: sharing_configs/admin.py [hunk 11]
```diff
@@ -187,25 +186,29 @@ def import_from_view(self, request, extra_context=None):
...
                     )
                     self.message_user(request, msg, level=messages.SUCCESS)
                     return redirect(reverse(main_url))
 
-                except ApiException as e:
+                except ApiException:
+
                     msg = format_html(
-                        _(f"Import of object failed: {e}"),
+                        _("Import of object failed"),
```
Comment (joeribekker) [inline sharing_configs/admin.py:206]:
```suggestion
                        _("The import of the selected item failed."),
```

File: testapp/admin.py [hunk 12]
```diff
@@ -1,34 +1,38 @@
...
     sharing_configs_import_form = ImportForm
 
     def get_sharing_configs_export_data(self, obj: object) -> bytes:
-        """return  django user model object as a byte like object"""
-
-        user = get_object_or_404(User, id=obj.id)
-        user_dict = model_to_dict(user)
-        dump_json_user = json.dumps(user_dict, sort_keys=True, default=str)
-        return dump_json_user.encode("utf-8")
+        """convert Theme model object to a byte like object"""
```
Comment (joeribekker) [inline testapp/admin.py:20]: Adjust comments to that in the example.

File: testapp/admin.py [hunk 13]
```diff
@@ -1,34 +1,38 @@
...
+        theme_dict = model_to_dict(theme)
+        theme_dict.pop("id", None)
+        dump_json_theme = json.dumps(theme_dict, sort_keys=True, default=str)
+        return dump_json_theme.encode("utf-8")
 
     def get_sharing_configs_import_data(self, content: bytes) -> object:
-        """return object from a byte string"""
-        obj = content.decode("utf-8")
-        return obj
+        """decode byte string and create an object of Theme model
```
Comment (joeribekker) [inline testapp/admin.py:28]: Also adjust comments to example

File: testapp/context_processors.py [hunk 14]
```diff
@@ -0,0 +1,20 @@
+from .models import Configuration, Theme
+
+
+def set_admin_color(request: object) -> dict:
+    """
+    return dict of color options
+    """
+    conf = Configuration.get_solo()
```
Comment (joeribekker) [inline testapp/context_processors.py:8]: See comments in example.

File: testapp/models.py [hunk 15]
```diff
@@ -0,0 +1,28 @@
...
+from django.utils.translation import gettext_lazy as _
+
+from solo.models import SingletonModel
+
+
+class Configuration(SingletonModel):
+    """
+    configuration for export/import admin color theme
+    """
+
```
Comment (joeribekker) [inline testapp/models.py:11]: See comments in example.

File: testapp/models.py [hunk 16]
```diff
@@ -0,0 +1,28 @@
...
+
+class Theme(models.Model):
+    """object to change color theme in admin"""
+
+    name = models.CharField(_("theme_name"), max_length=120)
+    primary = models.CharField(_("primary color"), max_length=60)
+    secondary = models.CharField(_("secondary color"), max_length=60)
+    accent = models.CharField(_("accent color"), max_length=60)
+    primary_fg = models.CharField(_("primary_fg"), max_length=60)
```
Comment (joeribekker) [inline testapp/models.py:25]: Can you use actually usefull labels? Repeating the attribute name is not usefull. At least remove the ``_``

File: testapp/templates/admin/base_site.html [hunk 17]
```diff
@@ -0,0 +1,9 @@
+{% extends "admin/base_site.html" %}
+{% load i18n static %}
+
+{% block extrastyle %}
+<html
+    style=" --primary:{{primary}}; --secondary:{{secondary}}; --accent:{{accent}}; --primary_fg:{{primary_fg}}; --selected-row:{{current_model_color}}">
```
Comment (joeribekker) [inline testapp/templates/admin/base_site.html:6]: Can you add an enter after each ``;``?

**Review by joeribekker — CHANGES_REQUESTED** [review]

File: README.rst [hunk 18]
```diff
@@ -87,4 +87,113 @@ the exporting file and override ``get_sharing_configs_export_data`` method
...
+    from django.utils.translation import gettext_lazy as _
+    from solo.models import SingletonModel
+
+
+    class Configuration(SingletonModel):
+        """
+        configuration for export/import of color theme in the Django admin
+        """
+
+        theme = models.ForeignKey("Theme", on_delete=models.CASCADE, related_name="theme")
```
[the remark this answers is not in the document]
Comment (frekkenSnork) [reply README.rst:110]: Adjusted

File: README.rst (hunk 3 above)
Comment (frekkenSnork) [reply README.rst:107]: yes

File: README.rst [hunk 19]
```diff
@@ -87,4 +87,113 @@ the exporting file and override ``get_sharing_configs_export_data`` method
...
+
+        def __str__(self) -> str:
+            return self.theme.name
```
[the remark this answers is not in the document]
Comment (frekkenSnork) [reply README.rst:113]: Done

File: README.rst [hunk 20]
```diff
@@ -87,4 +87,113 @@ the exporting file and override ``get_sharing_configs_export_data`` method
...
+
+Example
+=======
+Simple example for changing color theme in the django admin:
+
+Create two models: Configuration and Theme
+
+.. code-block:: python
+
+    from django.db import models
```
[the remark this answers is not in the document]
Comment (frekkenSnork) [reply README.rst:100]: Added

File: README.rst [hunk 21]
```diff
@@ -87,4 +87,113 @@ the exporting file and override ``get_sharing_configs_export_data`` method
...
+        accent = models.CharField(_("accent color"), max_length=60)
+        primary_fg = models.CharField(_("primary_fg"), max_length=60)
+        
+        
+Add two mixins provided by the sharing-configs app to the ThemeAdmin. 
+Thus you will get an interface to export and import ColorTheme objects to and from the API.  
+
+Customize two mixins methods: **get_sharing_configs_export_data** to convert model object to bytes and 
+**get_sharing_configs_import_data** to convert bytes to an object.
+Register both models with the admin site.
```
[the remark this answers is not in the document]
Comment (frekkenSnork) [reply README.rst:131]: Done

File: README.rst [hunk 22]
```diff
@@ -87,4 +87,113 @@ the exporting file and override ``get_sharing_configs_export_data`` method
...
+
+
+.. code-block:: python
+
+    import json
```
[the remark this answers is not in the document]
Comment (frekkenSnork) [reply README.rst:136]: Added

File: README.rst [hunk 23]
```diff
@@ -87,4 +87,113 @@ the exporting file and override ``get_sharing_configs_export_data`` method
...
+    from sharing_configs.admin import SharingConfigsExportMixin, SharingConfigsImportMixin
+    from sharing_configs.forms import ExportToForm, ImportForm
+
+    from .models import Configuration, Theme
+
+
+    class ThemeAdmin(SharingConfigsExportMixin, SharingConfigsImportMixin, admin.ModelAdmin):        
+
+        def get_sharing_configs_export_data(self, obj: object) -> bytes:
+            """return  django theme model object as a byte like object"""
```
[the remark this answers is not in the document]
Comment (frekkenSnork) [reply README.rst:151]: Done

File: README.rst [hunk 24]
```diff
@@ -87,4 +87,113 @@ the exporting file and override ``get_sharing_configs_export_data`` method
...
+            theme = get_object_or_404(Theme, id=obj.id)
+            theme_dict = model_to_dict(theme)
+            theme_dict.pop("id", None)
+            dump_json_theme = json.dumps(cleaned_theme_dict, sort_keys=True, default=str)        
+            return dump_json_theme.encode("utf-8")
+
+        def get_sharing_configs_import_data(self, content: bytes) -> object:
+            """
+            convert byte string into a dictionary and create a color-theme object;           
+            example api response: b'{"accent": "#f8f8f8", "name": "some-name", "primary": "#ab8585", "primary_fg": "#1a2b3c", "secondary": "#315980"}'
```
[the remark this answers is not in the document]
Comment (frekkenSnork) [reply README.rst:161]: Done

File: README.rst (hunk 5 above)
Comment (frekkenSnork) [reply README.rst:189]: Agree

File: README.rst [hunk 25]
```diff
@@ -87,4 +87,113 @@ the exporting file and override ``get_sharing_configs_export_data`` method
...
+Pass created variables to the ``base_site.html`` in the templates.
+.. code-block::
+  {% extends "admin/base_site.html" %}
+    {% block extrastyle %}
+    <html
+        style=" --primary:{{primary}}; --secondary:{{secondary}}; --accent:{{accent}}; --primary_fg:{{primary_fg}};">
+    </html>
+  {% endblock %}
+
+Now you can choose an available color theme via related congiguration object on its admin page.
```
[the remark this answers is not in the document]
Comment (frekkenSnork) [reply README.rst:199]: Changed

[the remark this answers is not in the document]
Comment (frekkenSnork) [reply README.rst:199]: Done

File: sharing_configs/utils.py (hunk 6 above)
Comment (frekkenSnork) [reply sharing_configs/utils.py:22]: Done

File: sharing_configs/client_util.py (hunk 7 above)
Comment (frekkenSnork) [reply sharing_configs/client_util.py:85]: Changes (also for files)

File: sharing_configs/client_util.py (hunk 8 above)
Comment (frekkenSnork) [reply sharing_configs/client_util.py:65]: Changed

File: sharing_configs/client_util.py (hunk 9 above)
Comment (frekkenSnork) [reply sharing_configs/client_util.py:50]: Done

File: sharing_configs/admin.py (hunk 10 above)
Comment (frekkenSnork) [reply sharing_configs/admin.py:197]: Changed

File: sharing_configs/admin.py (hunk 11 above)
Comment (frekkenSnork) [reply sharing_configs/admin.py:206]: Done

File: testapp/admin.py (hunk 12 above)
Comment (frekkenSnork) [reply testapp/admin.py:20]: Done

File: testapp/admin.py (hunk 13 above)
Comment (frekkenSnork) [reply testapp/admin.py:28]: Changed

File: testapp/templates/admin/base_site.html (hunk 17 above)
Comment (frekkenSnork) [reply testapp/templates/admin/base_site.html:6]: Done

File: testapp/context_processors.py (hunk 14 above)
Comment (frekkenSnork) [reply testapp/context_processors.py:8]: Done

File: testapp/models.py (hunk 15 above)
Comment (frekkenSnork) [reply testapp/models.py:11]: Changed

File: testapp/models.py (hunk 16 above)
Comment (frekkenSnork) [reply testapp/models.py:25]: Done

File: README.rst (hunk 4 above)
Comment (frekkenSnork) [reply README.rst]: Done

File: README.rst [hunk 26]
```diff
@@ -87,4 +87,113 @@ the exporting file and override ``get_sharing_configs_export_data`` method
...
+        accent = color_theme_obj.accent
+        primary_fg = color_theme_obj.primary_fg    
+        return {
+            "primary": primary_color,
+            "secondary": secondary_color,            
+            "primary_fg": primary_fg,
+            "accent": accent,
+            
+        }
+Pass created variables to the ``base_site.html`` in the templates.
```
[the remark this answers is not in the document]
Comment (frekkenSnork) [reply README.rst]: Done

File: testapp/models.py [hunk 27]
```diff
@@ -0,0 +1,28 @@
...
+
+
+class Configuration(SingletonModel):
+    """
+    configuration that holds the current theme
+    """
+
+    theme = models.ForeignKey("Theme", on_delete=models.CASCADE, related_name="theme")
+
+    def __str__(self) -> str:
```
Comment (joeribekker) [inline testapp/models.py:14]: Remove please, like in the example comments.

File: testapp/models.py [hunk 28]
```diff
@@ -0,0 +1,28 @@
...
+
+
+class Theme(models.Model):
+    """All attributes used for theming."""
+
+    name = models.CharField(_("name"), max_length=120)
+    primary = models.CharField(_("primary"), max_length=60)
+    secondary = models.CharField(_("secondary"), max_length=60)
+    accent = models.CharField(_("accent"), max_length=60)
+    primary_fg = models.CharField(_("primary_fg"), max_length=60)
```
Comment (joeribekker) [inline testapp/models.py:25]: Still labels with an underscore

File: README.rst [hunk 29]
```diff
@@ -87,4 +87,110 @@ the exporting file and override ``get_sharing_configs_export_data`` method
...
+            "secondary": theme.secondary,            
+            "primary_fg": theme.primary_fg,
+            "accent": theme.accent
+            
+        }
+
+Finally, pass the theme context variables to an overriden ``base_site.html`` in the templates folder.
+
+.. code-block::
+  {% extends "admin/base_site.html" %}
```
Comment (joeribekker) [inline README.rst:189]: This code block doesnt seem to render in markdown. Maybe indent is wrong or it needs a return after the block statement.

**Review by joeribekker — CHANGES_REQUESTED** [review]

**Review by joeribekker — APPROVED** [review]

**Pull request merged**